Use the python used with test-module to run modules

willthames opened the pull request:
##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
hacking/test-module

##### ANSIBLE VERSION
```
ansible 2.3.0 (devel 8c5d321a23) last updated 2016/12/21 11:33:16 (GMT +1000)
  config file = 
  configured module search path = Default w/o overrides
```

##### SUMMARY
If a non-system python is used to run test-module (as is
the case with brew python for OS X or in a virtualenv)
then that same python should be used when exploding
and executing the module.

willthames committed:
Use ansible_python_interpreter to run modules

Use ansible_python_interpreter to run modules if
`-I ansible_python_interpreter` is set.

Remove unused default from `-I` help text.

sivel commented:
test-module already supports supplying an interpreter path.  With this change wouldn't it break that functionality?  Also what if you are testing a non-python module?

bcoca commented:
its modifying ansiballz_setup, which should already only support 'new' style python modules

bcoca commented:
the part modifying run_test on the other hand....

willthames commented:
The interpreter path stuff doesn't solve this problem - it only changes the shebang for the wrapper ansiballz module. When the module gets extracted from the zip, it still has the /usr/bin/python shebang.

One could argue that it's ansiballz at fault here, or just ansible for insisting on /usr/bin/python shebangs on all (python) modules, and fix the issue at that source.

I hadn't considered other module types, I must confess. 

If there's a better fix, I'm all for it, but right now test-module doesn't work on OS X with brew python, and presumably it doesn't work in virtualenvs either.

willthames commented:
Virtualenv:

## Before fix
```
(boto) [wthames@wthames ansible (devel)]$ which python
~/venvs/boto/bin/python
(boto) [wthames@wthames ansible (devel)]$ python -c 'import boto'
(boto) [wthames@wthames ansible (devel)]$ hacking/test-module -I ansible_python_interpreter=`which python` -m lib/ansible/modules/cloud/amazon/rds.py -a 'command=facts instance_name=bobbins' 
* including generated source, if any, saving to: /home/wthames/.ansible_module_generated
* ansiballz module detected; extracted module source to: /home/wthames/debug_dir
***********************************
RAW OUTPUT

{"msg": "boto required for this module", "failed": true, ... }
```

## After fix
```
$ hacking/test-module -I ansible_python_interpreter=`which python` -m lib/ansible/modules/cloud/amazon/rds.py -a 'command=facts instance_name=bobbins'
* including generated source, if any, saving to: /home/wthames/.ansible_module_generated
* ansiballz module detected; extracted module source to: /home/wthames/debug_dir
***********************************
RAW OUTPUT

{"msg": "Region not specified. Unable to determine region from EC2_REGION.", "failed": true, ... }
```
(I'm trying to check if boto is installed, I haven't set up boto, so the latter error is the one I'd expect)

The nice thing is that you don't even *need* to set `-I` after the change (with ansiballz modules, I'm not sure the `-I` has any effect in either circumstance). Running the module under the same python interpreter as the one I'm using for everything else (i.e. `/usr/bin/env python`) is exactly what most people expect.

bcoca commented:
Except `env python` is not always 'the correct python' which is the reason we have ansible_python_interpreter in the first place.

willthames commented:
@bcoca, ansible_python_interpreter makes sense when running an ansible module against various target VMs rather than when running hacking/test-module locally (as at that point, `which python` is likely to be the python you want to use - and if not you can fix that!)

abadger commented:
It looks like this is the right level to make the fix but the fix needs to take its value from options.interpreter rather than hardcoding sys.executable.  It would also only add the interpreter if options.interpreter is given, otherwise use the present code path which uses the shebang from the file.

The reason that AnsiBallZ doesn't write a new shebang to the module is that would invalidate the module construction cache.  So the wrapper explicitly invokes the interpreter on the wrapped module.  This should (and does) take the same strategy of being explicit about the interpreter that invokes the module file.

abadger commented:
needs_revision

willthames commented:
Pushed a new commit that implementes @abadger's suggestion.

It's annoying that the default for `-I` is declared to be `python=sys.executable`  - if you were happy for this to work as documented this PR would be a lot simpler. However, I've chosen to remove the erroneous help text and use a python executable only if `-I ansible_python_interpreter` is set.

willthames committed:
Update test-module to pep8 standards

jctanner commented:
!needs_revision

abadger commented:
Merged to devel.  Thanks @willthames !
